lib/remap-lang/src/type_def.rs

```diff
@@ -35,37 +36,32 @@ pub struct TypeDef {
     ///     TypeDef {
     ///         fallible: false,
     ///         kind: Kind::Array,
-    ///         inner_type_def: Some(TypeDef {
+    ///         inner_type_def: InnerTypeDef::Array(TypeDef {
     ///             fallible: false,
     ///             kind: Kind::Null | Kind::Boolean,
-    ///             inner_type_def: None,
+    ///             inner_type_def: InnerTypeDef::None,
     ///         }.boxed()),
     ///     },
     /// );
     /// ```
-    pub inner_type_def: Option<Box<TypeDef>>,
+    pub inner_type_def: InnerTypeDef,
```

Comment:
> Any reason this got changed from an `Option<T>` to `T` with a variant `None`?

Mainly because it makes the `BitOr` and `BitAnd` operations simpler. 

We would then want to do `impl BitOr for Option<InnerTypeDef> {`, but of course, this doesn't work. Granted we could just extract that into a function that could be called instead of `|`, but that feels less elegant.

Reply:
Hmm. I see. I agree we shouldn't add our own functions for bitwise operations.

Still though, I'm guessing most of the bitwise operations happen on the top-level `TypeDef`, so we only have to deal with the awkward "bitwise two optionals" internally in `TypeDef`'s `BitOr` (etc) implementation, right?

If that's the case, I'd still prefer we take that approach, but if it causes too much complexity, I guess we can use this as the lesser of two evils 😄

Reply:
Ok, it's doable, but only if we lose the `BitAnd` operator since there is no correct answer to `Array & Map` that doesn't resolve to `None`. It's not used anyway so shouldn't be an issue.

Reply:
I have kept the `BitAnd` of TypeDef, and just moved all the `InnerTypeDef` functionality into that so it can handle the Nothing cases. Even though `BitAnd` is not used anywhere, I can see it coming in handy further down the line.